### DISCUSSION TITLE
Why no functions to read digital/analog values directly?

### DISCUSSION THREAD

**Discussion by chconnor** [MrYsLab/telemetrix#51] [Q&A] [answered]
Hi -- I'm excited to start using this. I'm a beginner, so I ask this question with humility: I was confused that there (apparently) are no functions to simply read the analog or digital value of a pin, and that the callback/reporting method is the only way to determine the values. Why is that?

For my project I can use the callback pattern, but it makes the coding significantly more cumbersome, and it makes me wonder if I'm missing something deeper about all this. E.g. if I want to check the value of an analog sensor once every 5 minutes, with maximal accuracy, it seems odd to be sending a stream of constant data over the serial port and running the callback method constantly (especially in a Python context where its single-threaded environment means you'd be interrupting any time-sensitive process, if I understand correctly.)

Any insight is appreciated, thanks!

**Comment by MrYsLab** [comment, marked as answer]
@chconnor Hi Casey,
   Those are all great questions. So why doesn't telemetrix provide a way to poll the current value of analog and digital inputs? The answer is that it is based on my experience working with Arduinos. That doesn't mean that there are situations that polling is a better solution, but I think those situations are few and far between. 

   In general, using callbacks makes things more efficient and usually simpler. For your example of wanting to "poll" an input every five minutes without being bombarded with reports you are not interested in, is to use the [disable_analog_reporting ](https://htmlpreview.github.io/?https://github.com/MrYsLab/telemetrix/blob/master/html/telemetrix/index.html#telemetrix.telemetrix.Telemetrix.disable_analog_reporting)method to turn reporting off, and then in five minutes turn it back on using [enable_analog_reporting](https://htmlpreview.github.io/?https://github.com/MrYsLab/telemetrix/blob/master/html/telemetrix/index.html#telemetrix.telemetrix.Telemetrix.enable_analog_reporting). 

   One caveat using this method is that reports are only generated when the current value read differs from the previous value. So if you do not get an immediate notification by turning reports back on, you can assume the value has not changed.

   Please let me know if I have answered all your questions, and please feel free to post any additional questions you might have.

Thanks,
Alan

**Comment by chconnor** [comment]
Thanks for the response, Alan! So I would need to set a boolean flag to false (which gets set to true in the callback), enable reporting, wait until the boolean goes true or a timeout expires, disable reporting, then check the value, right?

I understand the reasoning, but maybe you can consider this a humble feature request for the future. It'd be great to just use a single function call for reads when that's all we need. I came to the Telemetrix solution because the 8-pin Numato board I got can't notify on pin changes and can only do direct reads, and now with this solution it can only notify on changes and not do direct reads. :-)

Side question: when the reporting is determining whether a value has changed, is it comparing it to the most recent value of the pin, or the most recent value of the pin while reporting was enabled? E.g. if it's value 10 for a long time, then I disable reporting, then the value changes a lot, then it stabilizes at 10 again, then I enable reporting, do I get a report?

**Reply by MrYsLab** [reply]
Yes, you would need to implement a non-blocking timer. Using threads or asyncio would help provide that.

In terms of reporting, the current value is compared with the last active read. So if you turn off reporting, the last value recorded before reporting was disabled is compared to the current value when reporting is turned back on. In the scenario you describe, you would not get a new report. 

BTW, there is another feature that might be helpful to you. When you set the pin mode to analog input, a parameter allows you to set a differential threshold. The threshold value is handled as an absolute value, so a report is generated if the currently read value exceeds, plus or minus, the threshold value. 

I have no plans to implement a polling capability. It causes more problems than it solves. Most applications want to be aware of any changes. With polling, it is possible to miss value changes. In addition, it wastes value bandwidth on the serial port. However, if it is required for your application, you can always fork this repository and  Telemetrix4Arduino and make a custom version of the library.

**Reply by remcycles** [reply]
Thank you very much for all the work you've put into this project of making an improved Firmata.  It's very useful and saves me a ton of work.  I switched to Telemetrix because none of the active Python Firmata clients support I2C.

I'm glad @chconnor asked this when he did, right as I started learning about Telemetrix, because I had the same question.  I hacked  a fork to add digital and analog polling here:
https://github.com/remcycles/telemetrix/tree/polling
https://github.com/remcycles/Telemetrix4Arduino/tree/polling

It's pretty basic and not rigorously tested yet.

> Most applications want to be aware of any changes.

Maybe, but certainly not all.  Event-based programming is great, but the inversion of control that comes with it isn't always the best fit for simple applications.  It also doesn't match how most Arduino code is written, where directly reading pins is the norm.  It does seem like a good fit for an oscilloscope or DAQ type of system though.

In my use case I am doing hardware-in-the-loop testing, faking the system's inputs and reading it's outputs with an Arduino and Telemetrix.  My test code is driven by Behave on a desktop (already one sort of inversion of control there), and I'm essentially moving the I/O code from an Arduino .ino file to a desktop Python program.  So I'm looking for semantics similar to the Arduino I/O APIs.  I wouldn't be using pin interrupts in Arduino firmware and don't need them on the desktop.  I can wait for digital and analog signals to settle and I only need measure a DC value occasionally, not regularly sample an AC signal.

> In addition, it wastes value bandwidth on the serial port.

Both my application and @chconnor's once-every-five-minute sampling are low-bandwidth, so this is not an concern for me.  Besides, "enable reporting, I/O message, disable reporting" is more traffic than "read pin value, response".

> One caveat using this method is that reports are only generated when the current value read differs from the previous value. So if you do not get an immediate notification by turning reports back on, you can assume the value has not changed.

It might be helpful to send the current value of all pins when reporting is turned on so that callbacks are always called with the current value.  That way a user doesn't have to maintain their own cache of values and implement a timeout for callbacks that don't get called.

**Reply by MrYsLab** [reply]
Nice job. You clearly understand how to extend Telemetrix. I am going to offer an alternative solution, which I think is simpler to implement, requires no changes to either telemetirx or Telemetrix4Arduino, and is non-blocking.

Perhaps it is not apparent, but you do not need to implement a separate callback function for each pin you wish to monitor. You could create a single callback function to handle all digital and analog pin reports. You would register that single callback when setting the pin mode for each digital and analog pin. The callback function would use the first byte of its data parameter to determine the type of pin that is reporting. So, for example, for an analog input pin, data[0] would equal three, and data[1] identifies the PIN. For a digital pin, data[0] would equal two, and the PIN is contained in data[1].
In both cases, the reported value is contained in data[2]. You could enter the current reported value if you have a different dictionary for analog and digital pins, using the pin number as a key. Then, when you wish to do a read, you look up the pin in the appropriate dictionary.

Of course, you could create as many callbacks as you wish. One for each pin, one for all analog and digital reports, or even a single callback function to handle all reports.

I do not pretend to know your application better than you, but you might wish to consider the suggestion above.

Again, excellent job with the extension.

**Reply by remcycles** [reply]
Thanks.  You made it very easy to extend.  This was a good practice project for extending Telemetrix, and I'm confident I could add firmware support for more complicated devices if needed.

A dictionary cache was my first idea too, but I gave up on it because I wanted the polling semantics more.  Here's what I drafted:
```python
class Telemetix_Input_Cache:
    def __init__(self):
        self._cache = dict()
    def write(self, pin_type, pin_number, pin_value, raw_time_stamp):
        self._cache[pin_number] = (pin_type, pin_value, raw_time_stamp)
    def read(self, pin_number):
        try:
            return self._cache[pin_number][1]
        except KeyError:
            raise Exception(f'Have not seen an update from pin {pin_number}')

cache = Telemetix_Input_Cache()
cache.write("input", 10, 1, 0)
cache.read(10)
cache.read(11)
```

Now that I think about it, it might be nice to have something like that as a default callback in Telemetrix if the user doesn't supply something else.  Then there could be methods in Telemetrix to read pin state from the cache transparently.

**Reply by remcycles** [reply]
I won't open a pull request for the polling messages, since I don't expect you to maintain that feature/hack.  I mostly wanted to share the code to show how it could be done for those who want it.

**Reply by MrYsLab** [reply]
NP. I am unsure if my point was clear. By storing incoming reports in the callback, you could implement a "read" function that would access the latest reported value. As a result, there would be no blocking, and the newest value would always be available.

**Reply by remcycles** [reply]
That was my intent, to use `cache.write()` as the callback function.  But I misread the docs and thought the callback function received four parameters, not just the `data` list.  I should have read this example more closely:
https://github.com/MrYsLab/telemetrix/blob/master/examples/digital_input.py#L37

This is better:
```python
class Telemetix_Input_Cache:
    def __init__(self):
        self._cache = dict()
    def write(self, data):
        pin_number = data[1]
        self._cache[pin_number] = data
    def read(self, pin_number):
        try:
            return self._cache[pin_number][2]
        except KeyError:
            raise Exception(f'Have not seen an update from pin {pin_number}')
```
